server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/GitMirror.java

```diff
@@ -333,4 +375,322 @@ private Git openGit(File workDir) throws Exception {
+                                         entry.getKey().substring(localPath().length());
+            final long contentLength = applyPathEdit(dirCache, inserter, convertedPath, entry.getValue(), null);
+            numBytes += contentLength;
+            if (numBytes > maxNumBytes) {
+                throw new MirrorException("mirror contains more than " + maxNumBytes + " byte(s)");
+            }
+        }
+    }
+
+    private static long applyPathEdit(DirCache dirCache, ObjectInserter inserter, String pathString,
+                                      Entry<?> entry, @Nullable byte[] oldContent)
+            throws JsonProcessingException {
+        switch (EntryType.guessFromPath(pathString)) {
+            case JSON:
+                final JsonNode oldJsonNode = oldContent != null ? Jackson.readTree(oldContent) : null;
+                final JsonNode newJsonNode = firstNonNull((JsonNode) entry.content(),
+                                                          JsonNodeFactory.instance.nullNode());
+
+                // Upsert only when the contents are really different.
+                if (!Objects.equals(newJsonNode, oldJsonNode)) {
+                    // Use InsertText to store the content in pretty format
+                    final String newContent = newJsonNode.toPrettyString() + '\n';
+                    applyPathEdit(dirCache, new InsertText(pathString, inserter, newContent));
+                    return newContent.length();
+                }
+                break;
+            case TEXT:
+                final String sanitizedOldText = oldContent != null ?
+                                                sanitizeText(new String(oldContent, UTF_8)) : null;
+                final String sanitizedNewText = sanitizeText(entry.contentAsText());
+                // Upsert only when the contents are really different.
+                if (!sanitizedNewText.equals(sanitizedOldText)) {
+                    applyPathEdit(dirCache, new InsertText(pathString, inserter, sanitizedNewText));
+                    return sanitizedNewText.length();
+                }
+                break;
+        }
+        return 0;
+    }
+
+    private static byte[] currentEntryContent(ObjectReader reader, TreeWalk treeWalk) throws IOException {
+        final ObjectId objectId = treeWalk.getObjectId(0);
+        return reader.open(objectId).getBytes();
+    }
+
+    private static void maybeEnterSubtree(
+            TreeWalk treeWalk, String remotePath,
+            String path, boolean findingMirrorStateFile) throws IOException {
+        // Enter if the directory is under the remote path.
+        // e.g.
+        // path == /foo/bar
+        // remotePath == /foo/
+        if (path.startsWith(remotePath)) {
+            if (findingMirrorStateFile) {
+                // The mirror state file isn't placed under a subtree.
+                return;
+            }
```

Comment:
Nit: How about adding a dedicated method for the mirror state file? For example:
```java
void maybeEnterMirrorStateFileTee(...) {
   // This will cover all valid paths for `.mirror_state.json`
   // path == /foo
   // remotePath == /foo/bar/
   // = true
   
   // path == /foo
   // remotePath == /foo/bar/
   // = true

   // path == /foo/bar
   // remotePath == /foo/
   // = false
   if (remotePath.startsWith(path + '/')) {
       reeWalk.enterSubtree();
   }
}
```